Samples/Display recently visited products/Resources/RecentlyVisitedService.cs

```diff
@@ -0,0 +1,91 @@
+﻿using Litium.Products;
+using Litium.Runtime.DependencyInjection;
+using Litium.Web.Models.Products;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+
+namespace Litium.Accelerator.Services
+{
+    [Service(ServiceType = typeof(IRecentlyVisitedService), Lifetime = DependencyLifetime.Transient)]
+    public class RecentlyVisitedService : IRecentlyVisitedService
+    {
+        private const int NumberOfItemsToDisplay = 4; // this could be made into a website setting
+        private const string Key = "LitiumRecentlyVisited";
+
+        private readonly ProductModelBuilder _productModelBuilder;
+        private readonly VariantService _variantService;
+        private readonly BaseProductService _baseProductService;
+
+        private List<Item> _items;
+
+        public List<Item> Items {
+            get {
+                if (_items == null)
+                {
+                    var cookie = HttpContext.Current.Request.Cookies[Key];
+                    _items = cookie == null ? new List<Item>() : JsonConvert.DeserializeObject<List<Item>>(cookie.Value);
+                }
+                return _items;
+            }
+            set { _items = value; }
+        }
+
+        public RecentlyVisitedService(ProductModelBuilder productModelBuilder,
+            VariantService variantService,
+            BaseProductService baseProductService)
+        {
+            _productModelBuilder = productModelBuilder;
+            _variantService = variantService;
+            _baseProductService = baseProductService;
+        }
+
+        public void Add(ProductModel item)
+        {
+            var productSystemId = item.UseVariantUrl ? item.SelectedVariant.SystemId : item.BaseProduct.SystemId;
+
+            Add(productSystemId);
+        }
+
+        public void Add(Guid productSystemId)
+        {
+            // remove any existing item
+            var existingItem = Items.Where(x => x.Id == productSystemId).FirstOrDefault();
+            if (existingItem != null)
+            {
+                Items.Remove(existingItem);
+            }
+
+            Items.Add(new Item { Id = productSystemId, LastVisited = DateTime.Now });
+            
+            // trim the list
+            if (Items.Count > NumberOfItemsToDisplay + 1)
+            {
+                Items.RemoveRange(0, Items.Count - (NumberOfItemsToDisplay + 1));
+            }
+
+            var cookie = new HttpCookie(Key) { Value = JsonConvert.SerializeObject(Items) };
+            HttpContext.Current.Response.Cookies.Add(cookie);
+        }
+
+        public List<ProductModel> Get()
+        {
+            return Items
+                .OrderByDescending(x => x.LastVisited)
+                .Skip(1) // we add before we get, so don't want to show the item we're currently visiting
```

Comment:
Not tested but how would this logic work on for example the startpage, would it not remove the last item viewed when i actually want it in the list?

Reply:
Yes, the logic assumes it's displayed on a product page. I could simplify it so it would show the current product as well, then it would work out of the box on any other page. Maybe that's preferable since it's only supposed to be a starting point.